### PR TITLE
feat: clean containers and volumes up after a build

### PR DESCRIPTION
**Pull request by cdrage**
feat: clean containers and volumes up after a build

### What does this PR do?

* Removes the container after a success or unsuccessful build
* Removes any volumes created as a result of running the build

Other fixes:
* Moves the await show error message to outside of the try / catch as
  using 'await' on it was blocking the 'finally' statement regarding
  cleaning up
* Moves the "create container options" outside of the try / catch as it
  is not an async statement / it is just a generator

### Screenshot / video of UI

https://github.com/containers/podman-desktop-extension-bootc/assets/6422176/e2c6cdf1-cf29-412a-8e90-aae29a1a55e8

### What issues does this PR fix or reference?

Closes https://github.com/containers/podman-desktop-extension-bootc/issues/48

### How to test this PR?

1. Start a build
2. Stop the container mid-build
3. See the error output from building it
4. See that the container has been removed / any volumes removed / it's
   been cleaned up.

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

### CODE REVIEW

File: src/container-utils.ts [hunk 1]
```diff
@@ -123,3 +126,68 @@ export async function removeContainerIfExists(engineId: string, container: strin
...
   }
 }
+
+// Get all volume names that match the name of the container
+async function getVolumesMatchingContainer(engineId: string, container: string): Promise<string[]> {
+  try {
+    // Volumes returns a list of volumes across all engines. Only get the volume that matches our engineId
+    const engineVolumes = await extensionApi.containerEngine.listVolumes();
+    if (!engineVolumes) {
+      throw new Error('No volume providers found');
```
Comment (deboer-tim) [inline src/container-utils.ts:136]: Technically no volumes found? (could be providers)

File: src/container-utils.ts [hunk 2]
```diff
@@ -123,3 +126,68 @@ export async function removeContainerIfExists(engineId: string, container: strin
...
+    }
+    // If none are found, just return an empty array / small warning
+    const volumes = engineVolumes.find(volume => volume.engineId === engineId);
+    if (!volumes) {
+      console.log('Ignoring removing volumes, no volume engine providers found for engineId: ', engineId);
```
Comment (deboer-tim) [inline src/container-utils.ts:141]: Likewise, this is just no volumes on the engine we care about.

File: src/container-utils.ts [hunk 3]
```diff
@@ -123,3 +126,68 @@ export async function removeContainerIfExists(engineId: string, container: strin
...
+        });
+      });
+    });
+
+    // Remove any duplicates that may have been added (same volume used multiple times / mounted).
+    volumeNames = [...new Set(volumeNames)];
+    return volumeNames;
+  } catch (e) {
+    console.log(e);
+    throw new Error('There was an error getting the volumes: ' + e);
```
Comment (deboer-tim) [inline src/container-utils.ts:164]: If we have an error getting volumes this means we wouldn't even try to remove the container. IMHO we should still try, i.e. either don't throw this or put a catch around line 174.

**Comment by cdrage** [comment]
@deboer-tim Updated based on the comments

File: src/container-utils.ts (hunk 3 above)
Comment (cdrage) [reply src/container-utils.ts:164]: I've updated the function so that even if we error getting the volumes, we'll still try and remove the container.

**Review by deboer-tim — APPROVED** [review]

**Pull request merged**